### ISSUE TITLE
refactor: standardise vertical_velocity on [Nz, Ny, Nx] shape (drop Nz+1 inconsistency)

### ISSUE THREAD

**Issue by jejjohnson** [jejjohnson/finitevolX#210]
## Summary

`vertical_velocity` returns a `[Nz+1, Ny, Nx]` array — every other staggered field in the library is `[Nz, Ny, Nx]` (or `[Ny, Nx]` in 2D).  This breaks shape composition with `Mask3D.w` (which is `[Nz, Ny, Nx]`) and is the only place in the codebase that introduces a second 3D shape category.  We should refactor `vertical_velocity` to return `[Nz, Ny, Nx]` and absorb the bottom and top boundary values into the existing ghost-cell convention.

This was originally flagged in a review comment on PR #208 (`Mask3D` restructure) — it's outside the scope of that PR but tracked here for a follow-up.

## Background

The horizontal staggering on Arakawa C-grids in `finitevolX` uses **same-shape storage with a one-cell ghost ring**: every staggered field (`T`, `U`, `V`, `X`) has shape `[Ny, Nx]` and the physical interior is `[1:-1, 1:-1]`.  The 3D extension repeats this trick at every z-level: `[Nz, Ny, Nx]` for `T`, `U`, `V`, `X`, with `k = 0` and `k = Nz - 1` as top/bottom ghost shells.  See [docs/cgrid_discretization.md](../blob/refactor/mask-submodule-restructure/docs/cgrid_discretization.md) for the rationale.

The vertical interfaces — where `w` lives — could follow the same convention.  And they do, in `Mask3D.w` (since the mask restructure in PR #208):

```python
# finitevolx/_src/mask/cartesian.py
w_np = pool_bool(hf, kernel=(2, 1, 1), threshold=3.0 / 4.0, direction="trailing")
# w[k, j, i] = top face of T[k, j, i]; same shape as h.
```

But `vertical_velocity` in `operators/diagnostics.py` uses a different convention: `[Nz+1, Ny, Nx]` with explicit boundary indices.

```python
# finitevolx/_src/operators/diagnostics.py
def vertical_velocity(u, v, dx, dy, dz, mask=None) -> Float[Array, "Nzp1 Ny Nx"]:
    Nz, Ny, Nx = u.shape
    ...
    w = jnp.zeros((Nz + 1, Ny, Nx), dtype=u.dtype)
    cumsum = jnp.cumsum(div_h, axis=0)
    w = w.at[1:, :, :].set(-cumsum * dz)
    return w  # w[0] = 0 (sea floor), w[Nz] = sea surface
```

## The inconsistency in practice

```python
mask = Mask3D.from_mask(h_mask)         # mask.w.shape == (Nz, Ny, Nx)
w = vertical_velocity(u, v, dx, dy, dz) # w.shape    == (Nz+1, Ny, Nx)

w * mask.w  # ❌ shape mismatch
```

You cannot directly mask the output of `vertical_velocity` with `Mask3D.w` — they live in different shape categories.  Same problem with `mask.h * w`, `w[k] - h[k]`, or any `vmap` over z-levels that consumes both.

## Why it ended up this way

`vertical_velocity` was added in #82, before the same-shape convention had been formalised across the 3D code paths.  Convention B (`[Nz+1, Ny, Nx]`) was the natural choice for that function in isolation:

- The bottom-boundary condition `w(sea floor) = 0` is the initial condition for the bottom-up integration `w[k+1] = w[k] - div_h[k] * dz` — having `w[0]` as an explicit index makes the recurrence read naturally.
- The top-boundary value `w[Nz]` is the kinematic surface velocity, useful as a check for closure (rigid lid: should be zero; free surface: should equal `dη/dt`).
- One degree of freedom each at the top and bottom boundary maps cleanly onto the `Nz+1` interfaces around `Nz` cells.

`Mask3D` came later (in PR #208) and chose convention A to compose with `mask.h / u / v` and to vmap cleanly over z-levels.

Both are defensible in isolation; the problem is having both in the same library.

## Tradeoffs (from the docs)

| Property | Convention A — `[Nz, Ny, Nx]` | Convention B — `[Nz+1, Ny, Nx]` |
|---|---|---|
| Broadcasts with `h` / `u` / `v` | ✅ | ❌ |
| Consistent with horizontal staggering | ✅ | ❌ |
| `vmap` over z-levels across fields | ✅ | ❌ (shape mismatch) |
| Number of data-layout categories to learn | 1 | 2 |
| Explicit bottom boundary index | ❌ (in ghost shell) | ✅ (`w[0]`) |
| Explicit top boundary index | ❌ (in ghost shell) | ✅ (`w[Nz]`) |
| Clean free-surface / rigid-lid BC | Awkward | Natural |

Convention A wins on 7/8 hygiene criteria; convention B wins on the one criterion that is genuinely useful for the integration recurrence.

## Proposed refactor

**Standardise on convention A (`[Nz, Ny, Nx]`) throughout the library.**  Refactor `vertical_velocity` to return `[Nz, Ny, Nx]` with `w[k]` interpreted as the *top face of T[k]* — matching `Mask3D.w` and the trailing-pad convention adopted in PR #208.

### Code change

Current:

```python
def vertical_velocity(u, v, dx, dy, dz, mask=None) -> Float[Array, "Nzp1 Ny Nx"]:
    Nz, Ny, Nx = u.shape
    # ... compute div_h ...
    w = jnp.zeros((Nz + 1, Ny, Nx), dtype=u.dtype)
    cumsum = jnp.cumsum(div_h, axis=0)
    w = w.at[1:, :, :].set(-cumsum * dz)
    return w
```

Refactored:

```python
def vertical_velocity(u, v, dx, dy, dz, mask=None) -> Float[Array, "Nz Ny Nx"]:
    """Vertical velocity at top z-faces (w[k] = top face of T[k]).

    Stored at the same shape as h / u / v — bottom-boundary condition
    w(sea floor) = 0 is encoded by the bottom ghost cell (k = 0) having
    zero horizontal divergence.  The top face of the topmost real cell
    is at w[Nz - 2]; w[Nz - 1] is the top ghost shell.
    """
    Nz, Ny, Nx = u.shape
    # ... compute div_h, same as before ...
    # Integrate top-of-cell for each k: w[k] = -sum_{k'=0..k} div_h[k'] * dz
    return -jnp.cumsum(div_h, axis=0) * dz
```

### Test changes

`tests/test_diagnostics.py::TestVerticalVelocity` currently asserts:

```python
assert w.shape == (grid3d.Nz + 1, grid3d.Ny, grid3d.Nx)
np.testing.assert_allclose(w[0], 0.0, atol=1e-10)  # bottom boundary
```

Become:

```python
assert w.shape == (grid3d.Nz, grid3d.Ny, grid3d.Nx)
# Bottom boundary is encoded in the bottom ghost: only the bottom
# ghost's divergence contributes to w[0], which is typically zero.
# Verify directly via the cumsum invariant rather than indexing the
# extra layer.
```

The `test_known_w_uniform_divergence` test currently uses `w[k+1, ...]` indexing.  Under the refactor it would just use `w[k, ...]` with one fewer level of meaningful output.

### Doc changes

- Remove the `[Nz+1, Ny, Nx]` references in `grid/base.py`, `grid/cartesian.py`, `grid/spherical.py` docstrings (currently tell users that W-fields live "in a separate `[Nz+1, Ny, Nx]` array outside this grid container").
- Update `docs/cgrid_discretization.md` "Vertical staggering in 3D" section to drop the "currently uses both" note once the refactor lands.
- Update the `vertical_velocity` API docs to describe the new convention.

## Out of scope for this issue

- Adding a `Difference3D.diff_z_*` family of vertical-staggering operators.  Those would also benefit from a single shape convention but are independent feature work.
- Whether `Mask3D.w` should keep its current trailing-pad direction (top face) or switch to leading-pad (bottom face).  The current PR #208 uses trailing-pad to match the rest of the mask module; the refactored `vertical_velocity` should match that choice.

## Acceptance criteria

- [ ] `vertical_velocity` returns `[Nz, Ny, Nx]`.
- [ ] `Mask3D.w * vertical_velocity(...)` works (same shape, broadcasting succeeds).
- [ ] `tests/test_diagnostics.py::TestVerticalVelocity` passes with updated assertions.
- [ ] Grid module docstrings (`grid/base.py`, `grid/cartesian.py`, `grid/spherical.py`) no longer reference `[Nz+1, Ny, Nx]`.
- [ ] `docs/cgrid_discretization.md` "Vertical staggering in 3D" section drops the "currently uses both" caveat.

## References

- [docs/cgrid_discretization.md — Vertical staggering in 3D](../blob/refactor/mask-submodule-restructure/docs/cgrid_discretization.md#vertical-staggering-in-3d) — full pedagogical write-up of the two conventions
- PR #208 — original review comment
- Issue #82 — original `vertical_velocity` feature
